### ISSUE TITLE
Under the development of the team, Flutter SDK Version Consistency Management.

### ISSUE THREAD

**Issue by Sunzxyong** [flutter/flutter#26220]
Under the development of the team. This dependency on downloading the Flutter SDK locally for each development cannot guarantee the consistency of the Flutter SDK version and automatic management. If the Flutter SDK version is inconsistent at development time. This may cause problems with the Api compatibility of the Dart layer and the inconsistency of the Flatter virtual machine.

**Comment by Sunzxyong** [comment]
@liyuqian

**Comment by kangwang1988** [comment]
Yeah, guys using flutter on the same project should use the same flutter repo. There are several ways to do this, for example:
a. All guys using stable1.0
b. Write a shell to verify that the flutter environment used by guys on the same flutter project have the same flutter version.
c. You can also fork one in your gitlab repo and all guys use it to develop flutter.
Currently, I think it's up to you.
@Sunzxyong

**Comment by zoechi** [comment]
Duplicate of #17464 ?

**Comment by liyuqian** [comment]
I think that it's a feature request. As @kangwang1988 mentioned, Alibaba has several solutions to achieve this. I think the original issue reporter @Sunzxyong also implemented a solution in their own team.

I'd encourage @Sunzxyong and @kangwang1988 to share their solutions (scripts) and see if we could integrate it into flutter tools. At least, I think we can write some official guide/document on how to solve this issue if we're not going to provide an official tool. CC @Hixie, @tvolkert, @sfshaza2

**Comment by kangwang1988** [comment]
Now, developers in our team use the same git repo of flutter(forked in inner gitlab), even the bin/cache part is tracked. So that we can strictly limit that we all working with the same flutter.

**Comment by Sunzxyong** [comment]
I think this will bring more maintenance costs, and invaded the Flutter SDK. It is too intrusive and can't be easily upgraded smoothly. But other than that, this does not ensure that other developers do not download other versions of the Flutter SDK.

Now, our team is using the management of flutter wrapper to manage the flutter sdk version.
We plan to extend the **Flutter Intellij Plugin** to automatically generate a flutterw script when creating a Flutter project. The flutterw script is similar to the gradlew script, it will download and manage consistent versions automatically. 

Finally, we hope that the official support for the automated management of the Flutter SDK version. Make the Flutter community even stronger, I will submit a pull request later if needed.
@liyuqian @kangwang1988

**Comment by kangwang1988** [comment]
@Sunzxyong 
We have two main considerations:
1. There are serveral patches in engine which can't be upstreamed, but must be applied in our flutter repo. For example: https://github.com/flutter/flutter/issues/22353
It will result in black screen in certain devices.
2. We must ensure that the flutter we're using is stable for business usage. So in our development workflow, the engine/flutter repo must be stable, we don't need to keep up with the latest one. In fact, the patches(a few but important) approach is easy to use.
It not only help us when upgrading, but also fixes issues we've found but unable to upstream.

**Comment by Sunzxyong** [comment]
@kangwang1988 

We can change the Flutter SDK or do some customization and maintain a stable version for the current business in a private gitlab repo, But this does not affect the management of the Flutter SDK version.
If the gitlab has a bugfix version of the Flutter SDK, some developers check out this version, and some developers don't. This will cause version inconsistencies.

If there are multiple Flutter App businesses, and they use different versions of the Flutter SDK. This requires each Flutter project to have its own version of the Flutter SDK, so flutterw script is the solution to this.

The flutterw script will check out the specified Flatter SDK version for each developer automatically. This ensures that every developer uses the same Flutter SDK version in the same Flutter project.

**Comment by kangwang1988** [comment]
In our practice, developers use the same flutter git repo and branch.

**Comment by liyuqian** [comment]
@devoncarew for Flutter IDE plugins, @jonahwilliams for flutter tools.

**Comment by danielRi** [comment]
I also see this as a requirement. If QA tests 2 weeks and says everything is fine, lets build a release version! The dev building the release version upgraded his local flutter since he built the test version, he potentially installed an untested breaking change. 
@kangwang1988 I really like option b, e.G. a fastlane plugin would suffice.
